### PR TITLE
feat(notification-center): Added LogEvent notification and integration.  

### PR DESCRIPTION
**Pull request by rashidsp**
## Summary
-  Introduced `LogEvent` notification in `notification_center.rb`.
- Integrated it in `batch_event_dispatcher.rb` and `forwarding_event_processor.rb`.
- Added `notification_count` in `notification_center.rb` and integrated into `Optimizely` for `notification types` _ACTIVATE_  and  _TRACK_.
## Test plan
- Added unit tests for `LogEvent` in `notification_center_spec.rb`, `batch_event_dispatcher_spec.rb` and `forwarding_event_processor_spec.rb`.
- Updated `Optimizely's` unit tests for `LogEvent` and `notification_count`.

### CODE REVIEW

File: lib/optimizely.rb [hunk 1]
```diff
@@ -258,11 +258,13 @@ def track(event_key, user_id, attributes = nil, event_tags = nil)
       @event_processor.process(user_event)
       @logger.log(Logger::INFO, "Tracking event '#{event_key}' for user '#{user_id}'.")
 
-      log_event = EventFactory.create_log_event(user_event, @logger)
-      @notification_center.send_notifications(
-        NotificationCenter::NOTIFICATION_TYPES[:TRACK],
-        event_key, user_id, attributes, event_tags, log_event
-      )
+      if @notification_center.notification_count(NotificationCenter::NOTIFICATION_TYPES[:TRACK]).positive?
```
Comment (mjc1283) [inline lib/optimizely.rb]: What do we gain by checking notification count, as opposed to just sending notifications regardless of count?

File: lib/optimizely/event/batch_event_processor.rb [hunk 2]
```diff
@@ -36,10 +36,11 @@ class BatchEventProcessor < EventProcessor
 
     def initialize(
       event_queue: SizedQueue.new(DEFAULT_QUEUE_CAPACITY),
-      event_dispatcher:,
+      event_dispatcher: Optimizely::EventDispatcher.new,
```
Comment (mjc1283) [inline lib/optimizely/event/batch_event_processor.rb]: Why are we adding a default value here?

File: lib/optimizely/event/batch_event_processor.rb [hunk 3]
```diff
@@ -144,6 +151,16 @@ def run
 
         add_to_batch(item) if item.is_a? Optimizely::UserEvent
       end
+    rescue SignalException
```
Comment (mjc1283) [inline lib/optimizely/event/batch_event_processor.rb]: This seems like an important change, but it isn't discussed in the summary. Could you explain this change? If this isn't related to LogEvent notification, then I think it should be in its own separate PR.

File: lib/optimizely.rb (hunk 1 above)
Comment (rashidsp) [reply lib/optimizely.rb]: This validates that user has already added the listener before sending the notification.
Please see java-sdk as a reference: https://github.com/optimizely/java-sdk/blob/master/core-api/src/main/java/com/optimizely/ab/Optimizely.java#L304

File: lib/optimizely/event/batch_event_processor.rb (hunk 2 above)
Comment (rashidsp) [reply lib/optimizely/event/batch_event_processor.rb]: This is not related to LogEvent, therefore moved to PR: https://github.com/optimizely/ruby-sdk/pull/200

File: lib/optimizely/event/batch_event_processor.rb (hunk 3 above)
Comment (rashidsp) [reply lib/optimizely/event/batch_event_processor.rb]: Moved changes which are not related to LogEvent into PR: https://github.com/optimizely/ruby-sdk/pull/200.
